File: src/main/java/com/flink/tutorials/java/chapter4_api/transformations/MapExample.java
package com.flink.tutorials.java.chapter4_api.transformations;

import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

/**
 * This class shows different way to implement a `map()` transformation:
 *    * lambda function
 *    * anonymous class
 *    * implement MapFunction class
 * */

public class MapExample {

    public static void main(String[] args) throws Exception {

        StreamExecutionEnvironment senv = StreamExecutionEnvironment.getExecutionEnvironment();


        DataStream<Integer> dataStream = senv.fromElements(1, 2, -3, 0, 5, -9, 8);

        // lambda function
        // 使用Lambda表达式
        DataStream<String> lambdaStream = dataStream
                .map(input -> "lambda input : " + input + ", output : " + (input * 2));
        lambdaStream.print();

        DataStream<String> functionDataStream = dataStream.map(new DoubleMapFunction());
        functionDataStream.print();

        // anonymous class
        // 匿名类
        DataStream<String> anonymousDataStream = dataStream.map(new MapFunction<Integer, String>() {
            @Override
            public String map(Integer input) throws Exception {
                return "anonymous function input : " + input + ", output : " + (input * 2);
            }
        });
        anonymousDataStream.print();

        senv.execute("basic map transformation");
    }

    // implement MapFunction class
    // Two generic types for MapFunction: Integer for input and String for output
    // 继承MapFunction
    // 第一个泛型Integer是输入类型，第二个泛型String是输出类型
    public static class DoubleMapFunction implements MapFunction<Integer, String> {
        @Override
        public String map(Integer input) {
            return "function input : " + input + ", output : " + (input * 2);
        }
    }
}

File: src/main/scala/com/flink/tutorials/scala/chapter4_api/transformations/MapExample.scala
package com.flink.tutorials.scala.api.transformations

import org.apache.flink.api.common.functions.RichMapFunction
import org.apache.flink.streaming.api.scala._

object MapExample {

  def main(args: Array[String]): Unit = {

    // 创建 Flink 执行环境
    val senv: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment

    val dataStream: DataStream[Int] = senv.fromElements(1, 2, -3, 0, 5, -9, 8)

    // 使用 => 构造Lambda表达式
    val lambda = dataStream.map ( input => ("lambda Input : " + input.toString + ", Output : " + (input * 2).toString) )

    // 使用 _ 构造Lambda表达式
    val lambda2 = dataStream.map { _ * 2 }

    // 继承RichMapFunction
    // 第一个泛型Int是输入类型，第二个String泛型是输出类型
    class DoubleMapFunction extends RichMapFunction[Int, String] {
      override def map(input: Int): String =
        ("overide map Input : " + input.toString + ", Output : " + (input * 2).toString)
    }

    val richFunctionDataStream = dataStream.map {new DoubleMapFunction()}

    // 匿名类
    val anonymousDataStream = dataStream.map {new RichMapFunction[Int, String] {
      override def map(input: Int): String = {
        ("overide map Input : " + input.toString + ", Output : " + (input * 2).toString)
      }
    }}

    senv.execute("basic map transformation")
  }

}
